File: 884.uncommon-words-from-two-sentences.cpp
/*
 * @lc app=leetcode id=884 lang=cpp
 *
 * [884] Uncommon Words from Two Sentences
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<string> uncommonFromSentences(string A, string B) {
    unordered_map<string, int> uncommon;
    countUncommon(A, uncommon);
    countUncommon(B, uncommon);
    vector<string> ret;
    for (auto& pair : uncommon) {
      if (pair.second == 1) {
        ret.push_back(pair.first);
      }
    }
#if 0
    for (auto& word : ret) {
      cout << "@@@ \"" << word << "\"" << endl;
    }
#endif
    return ret;
  }

 private:
  void countUncommon(const string& s, unordered_map<string, int>& uncommon) {
    int len = s.length();
    int last = 0;
    for (int i = 0; i < len; i++) {
      string word;
      if (s[i] == ' ') {
        word = s.substr(last, i - last);
      } else if (i == len - 1) {
        word = s.substr(last, i - last + 1);
      }

      if (word.empty()) {
        continue;
      }

      if (uncommon.count(word)) {
        uncommon[word]++;
      } else {
        uncommon[word] = 1;
      }
      last = i + 1;
    }
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<string> exp;

  // Example 1
  exp = {"sour", "sweet"};
  ASSERT_EQ(s.uncommonFromSentences("this apple is sweet",
                                    "this apple is sour") == exp,
            true);

  // Example 2
  exp = {"banana"};
  ASSERT_EQ(s.uncommonFromSentences("apple apple", "banana") == exp, true);

  return 0;
}

File: 884.uncommon-words-from-two-sentences.rs
/*
 * @lc app=leetcode id=884 lang=rust
 *
 * [884] Uncommon Words from Two Sentences
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn uncommon_from_sentences(s1: String, s2: String) -> Vec<String> {
        let mut map = std::collections::HashMap::new();

        s1.split(' ')
            .chain(s2.split(' '))
            .for_each(|s| *map.entry(s).or_insert(0) += 1);
        map.iter()
            .filter_map(|(key, val)| match *val == 1 {
                true => Some((*key).to_string()),
                false => None,
            })
            .collect::<Vec<_>>()
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::uncommon_from_sentences(
            "this apple is sweet".to_string(),
            "this apple is sour".to_string()
        ),
        vec!["sour".to_string(), "sweet".to_string()]
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::uncommon_from_sentences("apple apple".to_string(), "banana".to_string()),
        vec!["banana".to_string()]
    );
}
